block/bpf_io_filter.c

```diff
@@ -121,52 +128,108 @@ int io_filter_prog_attach(const union bpf_attr *attr, struct bpf_prog *prog)
 	if (disk == NULL)
 		return -ENXIO;
 
-	//TODO: change to array of programs to allow multiple programs to attach
-	rcu_assign_pointer(disk->prog, prog);
+	ret = mutex_lock_interruptible(&disk->io_filter_lock);
+	if (ret)
+		return ret;
+
+	old_array = io_filter_rcu_dereference_progs(disk);
+	if (old_array && bpf_prog_array_length(old_array) >= BPF_MAX_PROGS) {
+		ret = -E2BIG;
+		goto unlock;
+	}
+
+	ret = bpf_prog_array_copy(old_array, NULL, prog, &new_array);
+	if (ret < 0)
+		goto unlock;
+
+	rcu_assign_pointer(disk->progs, new_array);
+	bpf_prog_array_free(old_array);
 
-	return 0;
+unlock:
+	mutex_unlock(&disk->io_filter_lock);
+	return ret;
 }
 
 int io_filter_prog_detach(const union bpf_attr *attr)
 {
 	struct bpf_prog *prog;
 	struct gendisk *disk;
 	struct fd f;
+	struct bpf_prog_array *old_array;
+	struct bpf_prog_array *new_array;
+	int ret;
 
 	if (attr->attach_flags)
 		return -EINVAL;
 
+	/* increments prog refcnt */
 	prog = bpf_prog_get_type(attr->attach_bpf_fd,
 				 BPF_PROG_TYPE_IO_FILTER);
 
 	if (IS_ERR(prog))
 		return PTR_ERR(prog);
 
 	f = fdget(attr->target_fd);
-	if (!f.file)
-		return -EBADF;
+	if (!f.file) {
+		ret = -EBADF;
+		goto put;
+	}
 
 	disk  = I_BDEV(f.file->f_mapping->host)->bd_disk;
-	if (disk == NULL)
-		return -ENXIO;
+	if (disk == NULL) {
+		ret = -ENXIO;
+		goto put;
+	}
 
-	rcu_assign_pointer(disk->prog, NULL);
+	ret = mutex_lock_interruptible(&disk->io_filter_lock);
+	if (ret)
+		goto put;
 
-	bpf_prog_put(prog);
+	old_array = io_filter_rcu_dereference_progs(disk);
+	ret = bpf_prog_array_copy(old_array, prog, NULL, &new_array);
+	if (ret)
+		goto unlock;
 
-	return 0;
+	rcu_assign_pointer(disk->progs, new_array);
+	bpf_prog_array_free(old_array);
+	bpf_prog_put(prog);	/* put for detaching of program from dev */
+
+unlock:
+	mutex_unlock(&disk->io_filter_lock);
+put:
+	bpf_prog_put(prog);	/* put for bpf_prog_get_type */
+	return ret;
 }
 
 int io_filter_bpf_run(struct bio *bio)
 {
-	struct bpf_prog *prog;
-	int ret = 0;
+	int ret = 1;	/* allow io by default */
+
+	if (bio->bi_disk->progs)
+		ret = BPF_PROG_RUN_ARRAY(bio->bi_disk->progs, bio, BPF_PROG_RUN);
+
+	return ret;
+}
+
+void io_filter_bpf_free(struct gendisk *disk)
+{
+	struct bpf_prog_array_item *item;
+	struct bpf_prog_array *array;
+	int ret;
+
+	ret = mutex_lock_interruptible(&disk->io_filter_lock);
```

Comment:
I thought about doing that but then I saw that mutex_destroy was an empty function and didn't see how it prevented future access to the mutex. Does mutex_destroy do anything (in non-debug mode)?

Reply:
Well it doesn't do anything in non-debug mode because we shouldn't be reaching here if somebody has taken a lock. So, adding a mutex_destroy() call in the place where we intend to destroy the mutex will help us debug any situations where we wrongly take locks. So, I think it's better to call destroy in the caller and and then not worry about mutex here at all.

Reply:
done